include/wil/safearrays.h

```diff
@@ -0,0 +1,897 @@
+    //!     auto current = std::begin(source);
+    //!     for(BSTR& bstr : sa.access_data())
+    //!     {
+    //!         // Transfer ownership of the BSTR into the safearray
+    //!         bstr = (*current++).release();
+    //!     }
+    //!     return sa;
+    //! }
+    //! ~~~~
+    //! void handle_safearray_of_bools(SAFEARRAY* psa, std::function<void(bool, ULONG)> fnHandler)
+    //! {
+    //!     auto data = wil::safearraydata<VARIANT_BOOL>{psa};
+    //!     for (auto i = 0U; i < data.size(); ++i)
+    //!     {
+    //!         fnHandler(data[i] != VARIANT_FALSE, i);
+    //!     }
+    //! }
+    //! ~~~~
+    //! template<typename T>
+    //! std::vector<T> extract_all_values(SAFEARRAY* psa)
+    //! {
+    //!     auto data = wil::safearraydata<T>{psa};
+    //!     // Construct the vector from the iterators
+    //!     return {data.begin(), data.end()};
+    //! }
+    //! ~~~~
+    //! template<typename INTERFACE>
+    //! std::vector<wil::com_ptr<T>> extract_all_interfaces(SAFEARRAY* psa)
+    //! {
+    //!     auto result = std::vector<wil::com_ptr<INTERFACE>>{};
+    //!     auto data = wil::safearraydata<LPUNKNOWN>{psa};
+    //!     result.reserve(data.size());
+    //!     for(auto& p : data) // type of P is LPUNKNOWN
+    //!     {
+    //!         // Use "tag_com_query" if you want failure instead of nullptr if INTERFACE not supported
+    //!         result.emplace_back(p, details::tag_try_com_query);
+    //!     }
+    //!     return result;
+    //! }
+    //! ~~~~
+    //! @tparam T           The underlying datatype expected in the SAFEARRAY. It should be a scalar type, BSTR, VARIANT_BOOL, LPUNKNOWN/LPDISPATCH, etc.
+    //! @tparam err_policy  Represents the error policy for the class (error codes, exceptions, or fail fast; see @ref page_errors)
+    template <typename T, typename err_policy = err_exception_policy>
+    class safearraydata_t
+    {
+    private:
+        typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnaccessData), details::SafeArrayUnaccessData, details::pointer_access_noaddress> safearray_unaccess_data;
+
+    public:
+        typedef typename err_policy::result result;
+        typedef typename safearray_unaccess_data::pointer pointer;
+        typedef T value_type;
+        typedef value_type* value_pointer;
+        typedef const value_type* const_value_pointer;
+        typedef value_type& value_reference;
+        typedef const value_type& const_value_reference;
+        typedef value_type* iterator;
+        typedef const value_type* const_iterator;
+
+    public:
+        // Exception-based construction
+        safearraydata_t(pointer psa)
+        {
+            static_assert(wistd::is_same<void, result>::value, "this constructor requires exceptions; use the create method");
+            access(psa);
+        }
+        safearraydata_t() = default;
+        safearraydata_t(safearraydata_t&&) = default;
+        ~safearraydata_t() = default;

```

Comment:
Unnecessary

Reply:
Not sure to which line you're referring.  The move constructor is necessary for the `safearray_unaccess_data` member to move correctly.  The rest might not be strictly necessary, but they do show that I didn't forget about them, and certainly don't hurt anything.

Reply:
Sorry, I keep forgetting that Github's PR experience is a bit sub-optimal and doesn't show the highlighted range. I originally meant the destructor, however I realize now that none of the 5 (copy/move constructor/assignment and destructor) are necessary since that's the behavior of `unique_any<>`

Reply:
I think they are necessary.  `safearraydata_t` isn't a `unique_any` type because it needs additional data members (pointer and size).  It **has** a `unique_any` type that ensures `SafeArrayUnaccessData` is called properly and that requires the class to be "move only", so I defaulted the ones that are okay and deleted the ones that aren't.

Which do you think I should remove?

Reply:
Right. Members affect default constructors/assignment operators in much the same - if not the exact same - way as base classes. E.g. see https://godbolt.org/z/cb77s1. In general, you should either define (actually define, not default) all five or let the compiler do the correct thing for you. E.g. see [the rule of five](https://en.cppreference.com/w/cpp/language/rule_of_three)